ci/scripts/check-bazel-tags.sh

```diff
@@ -6,14 +6,119 @@
 
 set -e
 
-untagged=$(./bazelisk.sh query "rdeps(//..., //hw:verilator) except attr(tags, '[\\[ ](verilator|manual)[,\\]]', //...)")
-if [[ ${untagged} ]]; then # Check that all targets that depend on verilator are tagged
-  echo "Target(s): ${untagged} depend(s) on //hw:verilator, please tag it with verilator or manual";
-  exit 1
-fi
+# The list of bazel tags is represented as a string and checked with a regex
+# https://bazel.build/query/language#attr
+# This function takes a tag(or regex component) and wraps it so attr can query
+# for exact matches.
+exact_regex () {
+  echo "[\\[ ]${1}[,\\]]"
+}
 
-untagged=$(./bazelisk.sh query "rdeps(//..., kind('bitstream_splice', //...)) except attr(tags, '[\\[ ](cw310_rom|cw310_test_rom|vivado|manual)[,\\]]', //...)")
-if [[ ${untagged} ]]; then # Check that all targets that depend on vivado are tagged
-  echo "Target(s): ${untagged} depend(s) on a bitstream_splice but isn't tagged with vivado or manual";
+check_empty () {
+if [[ ${1} ]]; then
+  echo "Error:"
+  echo "$1"|sed 's/^/    /';
+  echo "$2"
   exit 1
 fi
+}
+
+# This check ensures OpenTitan software can be built with a wildcard without
+# waiting for Verilator using --build_tag_filters=-verilator
+untagged=$(./bazelisk.sh query \
+  "rdeps(
+      //...,
+      //hw:verilator
+  )
+  except
+  attr(
+      tags,
+      '$(exact_regex "(verilator|manual)")',
+      //...
+  )" \
+  --output=label_kind)
+check_empty "${untagged}" \
+"Target(s) above depend(s) on //hw:verilator, please tag it with verilator or
+(to prevent matching any wildcards) manual.
```

Comment:
I guess that comma could be a semicolon or a period...?

Reply:
BTW (I guess because the edit is on the previous line) this comment shows outdated code in the conversations tab but doesn't show as outdated.

Reply:
@dmcardle I don't know what you're asking for here so I'm blocked on resolving it.

Reply:
It seemed to me that this guidance could be a little more pointed.

Is the goal to ensure that users without Vivado can run `bazel build //...`? Then tagging these targets with 'manual' is mandatory.

Is the goal to ensure that CI can entirely exclude Vivado-related targets with the 'vivado' tag? Then, tagging with 'vivado' is mandatory.

If both are goals, we should instruct the user to add both tags, right?

Reply:
Sorry, I'm confused, I just realized this comment was about Verilator.

Reply:
No worries, the verilator and vivado tags are effectively similar:
Verilator has a longer set of precedents, a failure case that can be mitigated by building a dummy and consensus that intermediate deps shouldn't really be tagged.

There's also not a plan to mitigate the need for the verilator tag like there is for vivado.

The verilator case should be hard to hit: the functest rule adds the tag, so we shouldn't see new targets hit this at all, so the contents of this error message shouldn't be that relevant.

I'll feel free to resolve this and post it as an issue if this PR needs merging, but until then this PRs as good a place as any to sort it out.

The goal is to ensure users without vivado can run `bazel build //... --build_tag_filters=-vivado` With that invocation either manual or vivado will prevent matching and building. I don't like having manual targets with no rdeps because they're likely to reflect dead code. (test_suites are an exception here, kindof)
if the target has rdeps then we have been tagging it manual without objection in the verilator context, and could probably do the same for vivado.)

Other than the rdeps thing, there isn't a consensus and I don't have a notion as to which should be used.